src/lib/comms/sol-socket-dtls-impl-tinydtls.c

```diff
@@ -205,66 +207,43 @@ sol_socket_dtls_recvmsg(struct sol_socket *socket, void *buf, size_t len, struct
         return -EAGAIN;
     }
 
-    if (!buf)
-        return item->buffer.used;
-
     memcpy(cliaddr, &item->addr, sizeof(*cliaddr));
 
-    if (item->buffer.used <= len) {
-        memcpy(buf, item->buffer.data, item->buffer.used);
-        return remove_item_from_vector(&s->read.queue, item,
-            item->buffer.used);
-    }
-
-    memcpy(buf, item->buffer.data, len);
-
-    buf_copy = sol_util_memdup((const char *)item->buffer.data + len,
-        item->buffer.used - len);
-    SOL_NULL_CHECK_GOTO(buf_copy, clear_buf);
-
-    sol_buffer_init_flags(&new_buf, buf_copy, len,
-        SOL_BUFFER_FLAGS_CLEAR_MEMORY | SOL_BUFFER_FLAGS_NO_NUL_BYTE |
-        SOL_BUFFER_FLAGS_FIXED_CAPACITY);
-    new_buf.used = item->buffer.used - len;
-
-    sol_buffer_fini(&item->buffer);
-    item->buffer = new_buf;
-
-    return len;
+    ret = sol_buffer_set_buffer(buf, &item->buffer);
+    SOL_INT_CHECK(ret, < 0, ret);
 
-clear_buf:
-    SOL_WRN("Could not copy buffer for short read, discarding unencrypted data");
-    return remove_item_from_vector(&s->read.queue, item, -ENOMEM);
+    return remove_item_from_vector(&s->read.queue, item, buf->used);
 }
 
 static ssize_t
-sol_socket_dtls_sendmsg(struct sol_socket *socket, const void *buf, size_t len,
+sol_socket_dtls_sendmsg(struct sol_socket *socket, const struct sol_buffer *buf,
     const struct sol_network_link_addr *cliaddr)
 {
     struct sol_socket_dtls *s = (struct sol_socket_dtls *)socket;
     struct queue_item *item;
-    void *buf_copy;
+    int ret;
 
     if (s->write.queue.len > 4) {
         SOL_WRN("Transmission queue too long");
         return -ENOMEM;
     }
 
-    buf_copy = sol_util_memdup(buf, len);
-    SOL_NULL_CHECK(buf_copy, -ENOMEM);
-
     item = sol_vector_append(&s->write.queue);
     SOL_NULL_CHECK(item, -ENOMEM);
 
     item->addr = *cliaddr;
-    sol_buffer_init_flags(&item->buffer, buf_copy, len,
-        SOL_BUFFER_FLAGS_CLEAR_MEMORY | SOL_BUFFER_FLAGS_NO_NUL_BYTE |
-        SOL_BUFFER_FLAGS_FIXED_CAPACITY);
-    item->buffer.used = len;
+    sol_buffer_init(&item->buffer);
```

Comment:
why not keep the `CLEAR_MEMORY` and `NO_NUL_BYTE` flags? they do make sense to keep